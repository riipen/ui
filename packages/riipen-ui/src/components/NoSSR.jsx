import PropTypes from "prop-types";
import React from "react";

class NoSSR extends React.Component {
  static propTypes = {
    /**
     * The children not to render server side.
     */
    children: PropTypes.node,

    /**
     * The fallback content to display.
     */
    fallback: PropTypes.node
  };

  static defaultProps = {
    fallback: null
  };

  constructor(props) {
    super(props);

    this.state = {
      mounted: false
    };
  }

  componentDidMount() {
    // eslint-disable-next-line react/no-did-mount-set-state
    this.setState({
      mounted: true
    });
  }

  render() {
    const { children, fallback } = this.props;
    const { mounted } = this.state;

    // We need the Fragment here to force react-docgen to recognise NoSSR as a component.
    return <React.Fragment>{mounted ? children : fallback}</React.Fragment>;
  }
}

export default NoSSR;
