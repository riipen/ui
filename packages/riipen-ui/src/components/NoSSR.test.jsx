import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import NoSSR from "./NoSSR.jsx";

describe("<NoSSR>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<NoSSR />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<NoSSR />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default prop", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<NoSSR />);

      const component = wrapper.find("NoSSR");
      expect(component.props().fallback).toEqual(null);
    });
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <span>pizzas</span>;

      const wrapper = mount(<NoSSR>{child}</NoSSR>);

      expect(wrapper.contains(child)).toEqual(true);
    });

    it("renders children over fallback when mounted", () => {
      const child = <span>pizzas</span>;
      const fallback = <span>fallback</span>;

      const wrapper = mount(<NoSSR fallback={fallback}>{child}</NoSSR>);

      wrapper.setState({ mounted: true });

      expect(wrapper.contains(child)).toEqual(true);
      expect(wrapper.contains(fallback)).toEqual(false);
    });
  });

  describe("fallback prop", () => {
    it("renders fallback over children when mounted is false", () => {
      const child = <span>pizzas</span>;
      const fallback = <span>fallback</span>;

      const wrapper = mount(<NoSSR fallback={fallback}>{child}</NoSSR>);

      wrapper.setState({ mounted: false });

      expect(wrapper.contains(child)).toEqual(false);
      expect(wrapper.contains(fallback)).toEqual(true);
    });
  });
});
