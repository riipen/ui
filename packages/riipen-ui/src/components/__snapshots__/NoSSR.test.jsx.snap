// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoSSR> renders correct snapshot 1`] = `
<NoSSR
  fallback={null}
/>
`;
